### ISSUE TITLE
npm network timeout -- but I have no proxy and full network connectivity to the registry server

### ISSUE THREAD

**Issue by TimeForAChange** [npm/npm#8293]
System: Windows 7 64bit, using a "git bash" environment.

I refer to this closed bug: https://github.com/npm/npm/issues/7781

I'm using io.js v2.0.2 and I don't need a proxy, can ping registry.npmjs.org, `npm view npm version` shows 2.10.0. `npm install` results in a network timeout.

![unbenannt-1](https://cloud.githubusercontent.com/assets/199179/7670352/c42a09a6-fca1-11e4-9c8d-057f6116b9df.png)

npm debug log:

https://gist.github.com/hasenstein/66c9b9d024cc4a11ee1e

I can ping the host and any other download attempts I made never fail (git for example).

Command `$ curl https://registry.npmjs.org/amdefine` works fine, for example, while the exact same download fails for npm.

There is no proxy configured (which is correct, my consumer DSL device router only blocks incoming requests): according to `npm config ls -l | grep prox`

https-proxy = null
proxy = null

**Comment by hasenstein** [comment]
npm debug log:

https://gist.github.com/hasenstein/66c9b9d024cc4a11ee1e

**Comment by hasenstein** [comment]
`npm install express` worked!!!

But `npm install` (from package.json)  does not?

**Comment by othiym23** [comment]
It sounds like you may have a flaky network connection, because at the network level, installing from `package.json` and from the command line use the exact same underlying logic to talk to the registry.

**Comment by hasenstein** [comment]
No I don't. M<y network is reliable. I'm on the network at the same time as this happens. And this is 100% and reliably reproducible. It ONLY and ALWAYS happens on "npm install". I can repeat the experiment many times - either this or installing a package by name. One works, the other does not. It is not my network.

**Comment by hasenstein** [comment]
That process -- AND ONLY THAT PROCESS -- gets only SYN_SENT for https connections to 185.31.17.162. I can make such connections without problem from the command line or the browser at the same time this is happening (not getting anything back -- that seems to be a shared-by-name webserver that just returns "unknown domain" but that's beside the point).

![unbenannt-2](https://cloud.githubusercontent.com/assets/199179/7722840/23aa3b00-fee4-11e4-8b0f-c358ed9d9f8e.png)

**Comment by hasenstein** [comment]
And now suddenly, after changing nothing but running all these experiments, it works. After finding the `--verbose` option and testing its usefulness with the command above (last comment, install package "sinon") I now tested it with `npm --verbose install` and from here on out that command works.

**Comment by phillmill** [comment]
I'm having this exact problem, do you have any more insight on getting it to work?

**Comment by hasenstein** [comment]
Not really, I wrote everything I have done. All I did between "it doesn't work" and "it works" (within a few minutes, I was experimenting for this very bug report, to add info) was what I said above, use the `--verbose` option.

**Comment by phillmill** [comment]
I actually managed to solve it by setting my registry :S

npm set registry https://registry.npmjs.org/

**Comment by hasenstein** [comment]
Which is strange, given that it already tried to connect to the registry. Can't be the root cause.

**Comment by hasenstein** [comment]
I just tried `npm update` and again I get what the earlier screenshot above shows - lots of SYN_SENT. The `npm set registry...` suggestion made no difference -- not surprisingly, since it had been using the correct registry as is evident by the IP address destination of the SYN_SENT packets, which is that exact server.

**Comment by hasenstein** [comment]
I just tried removing io.js and installing node.js, and -- using npm version 2.10.1 -- I get the same result.

**Comment by atinsood** [comment]
Just a suggestion, if you are using third party software for allowing/blocking applications from accessing network, make sure that node is not in the block list of software.

**Comment by hasenstein** [comment]
I only use the Windows Firewall. I disabled my Anti-Virus (AVG Free - does not contain a firewall) and I also disabled Windows Firewall, just to have tried that too (it could not be the culprit). No change.

**Comment by AW-UC** [comment]
Same here on Windows 7 64bit and my network is decent. I guess installing individual packages is just less likely to fail, because only a single package is installed. With a `package.json` many packages are installed, so it is more likely that it will fail on one. Maybe running `npm install package` several times in a row, will finaly fail as well.
It might help to have an option to increase timeouts or retry attempts.

**Comment by hasenstein** [comment]
It does not seem to be related to probabilities. For me, individual installs _always_ work no matter how many. `npm install` requests _all_ fail, 100%.

**Comment by AW-UC** [comment]
I just experienced that issue with `npm install -g package` as well. @hasenstein what happens, if you install with `-g` flag?

**Comment by hasenstein** [comment]
I tried from my Windows 8.1 ultrabook - same result. Then I tried a different route: Instead of going through my home router - which is old but it has not caused me any connectivity problems ever - I used the 3G connection that I have on that ultrabook. It works. So it isn't the OS or the PC's network, it's the router. But it is something that only appears with this one issue here. And npm installation of single packages works.

**Comment by fightingtheboss** [comment]
I am experiencing this problem as well, running Vagrant on Ubuntu 14.04 LTS (Trusty Tahr). Inconsistent installation results when running `npm install` within a project with a `package.json`. If I install packages individually it seems to work.

My working theory is that npm times out / is blocked on downloads, for some reason, which then causes the rest of the installation to silently end. It's definitely not my network connection. It's making my dev setup really difficult to keep consistent. I plan to setup a test harness to reproduce this problem, but in the meantime if anyone has any Linux-based suggestions for monitoring npm traffic, I'd be up to try them out.

**Comment by fightingtheboss** [comment]
To clarify, I haven't seen actual timeout errors like the original poster, I just see the installation end prematurely and when I try to run the project, dependencies are missing.

**Comment by pkvpraveen** [comment]
Same problem here
Microsoft Windows [Version 6.1.7601]
Copyright (c) 2009 Microsoft Corporation. All rights reserved.

C:\Users\Praveen>npm -v
2.11.2

C:\Users\Praveen>npm install -g bower
npm ERR! Windows_NT 6.1.7601
npm ERR! argv "C:\Program Files\nodejs\\node.exe" "C:\Program Files\nodejs\node_modules\npm\bin\npm-cli.js" "install" "-g" "bower"
npm ERR! node v0.12.5
npm ERR! npm v2.11.2
npm ERR! code ETIMEDOUT
npm ERR! errno ETIMEDOUT
npm ERR! syscall connect

npm ERR! network connect ETIMEDOUT
npm ERR! network This is most likely not a problem with npm itself
npm ERR! network and is related to network connectivity.
npm ERR! network In most cases you are behind a proxy or have bad network settings.
npm ERR! network
npm ERR! network If you are behind a proxy, please make sure that the
npm ERR! network 'proxy' config is set properly. See: 'npm help config'

npm ERR! Please include the following file with any support request:
npm ERR! C:\Users\Praveen\npm-debug.log

**Comment by pkvpraveen** [comment]
ok, I had a firewall on my router(belkin) too. Turning it off fixed the problem.

**Comment by hasenstein** [comment]
What was the firewall blocking? Usually firewalls allow all outbound connections.

**Comment by sergiovilar** [comment]
Same problem here.

**Comment by sergiovilar** [comment]
@pkvpraveen I have a belkin router too, turned off the firewall and now works. Thanks!

**Comment by hasenstein** [comment]
I don't see how that firewall issue is related to the originally reported one (by me). Because while I do have a firewall it allows all outbound connections, and given my problem description it ISN'T a simple "firewall issue".

I would like to point that out so that it doesn't sound as if those responses solve anything. Or, alternatively, if it is related to something "firewally", wouldn't it be interesting to find out why *everything but this* works (even npm install partially works)?

**Comment by sergiovilar** [comment]
@hasenstein I was just searching about the problem and found the answer here. Sorry to flood the thread.

**Comment by ahouchens** [comment]
Just verified that turning off the belkin router firewall fixes this issue. Maybe try turning off your own router firewall or adjusting rules accordingly.

**Comment by hasenstein** [comment]
Does nobody listen (i.e. read)? You guys with firewall issues have a _different_ issue. "Fix your firewall" *already* is THE default answer for connection issues and rightfully so. I would prefer if those simple issues that are fixed by the standard answer stay out of this issue. Because *obviously* I don't have a firewall issue. See above. I *can* install stuff after all, not to mention that I could always reach the destination "manually".

**Comment by iarna** [comment]
@hasenstein So we have seen folks with networks that otherwise seem to work fine but explode when too many concurrent connections are made. I wonder if you wouldn't mind testing this… ?

To test this, give https://www.npmjs.com/package/crapify a try with concurrency set modesty low (say 5) but otherwise leaving the connection alone.

If that still fails then there's some other sort of gremlin at work. If that doesn't fail, well, you've got a work around and I may be able to at least point you in the right direction as far as root causes go…

**Comment by hasenstein** [comment]
When I made my report I didn't have such problems, but the last two days at least I often have trouble getting any connection at all to www.npmjs.com (via browser). ONLY this website.

http://status.npmjs.org/  and http://www.isitdownrightnow.com/npmjs.org.html show it should be up, but I got to https://www.npmjs.com/ only after the n-th attempt.

Anyway, I'll get a new router for security reasons, mine got the last vendor update 2011 and Google complains about strange requests from my site and I know it isn't the PC so my router may have been compromised.

**Comment by mitchellbusby** [comment]
Just wanted to add I had the same issue with the belkin router firewall, and it was fixed after disabling the firewall. I'm a bit hesitant to keep the firewall disabled; has anyone found a good exception rule to add to it?

**Comment by TimeForAChange** [comment]
That _should_ should not be necessary if your normal web browsing works. It's just port 80 HTTP traffic or 443 SSL. If it _is_ (necessary) something is really weird here. If only I wasn't so lazy to start debugging with Wireshark, and I have a brand-new router now anyway because my old one got hacked (Google complained and would not let me search any more).

**Comment by othiym23** [comment]
I am pretty sure that this is a duplicate of #7862, or at any rate is a manifestation of the same underlying problem. In order to keep all of the folks affected in the same place, I'm closing this as a duplicate of that issue. Please do keep an eye on that thread and contribute with your own information, because the CLI team is still trying to figure out enough of what's going on to figure out what to look at next.

**Comment by othiym23** [comment]
(And the only protocol I think that you'll use regularly aside from HTTP and HTTPS is git's wire protocol, used by some git-hosted dependencies, which uses TCP port 9418. I would be neither surprised nor unsurprised to learn that a consumer-grade Internet router's firewall blocked git by default.)

**Comment by aditya-bhuyan** [comment]
I also faced the same issue.

I actually managed to solve it by setting my registry 

npm set registry https://registry.npmjs.org/

**Comment by TimeForAChange** [comment]
> I would be neither surprised nor unsurprised to learn that a consumer-grade Internet router's firewall blocked git by default.

I have yet to find a consumer router that blocks any outgoing traffic. None of mine ever did. Outgoing TCP traffic should not be a problem.

Quite the opposite: The router makers want to avoid as much trouble arriving on their hotlines as possible, they have every incentive to block only what's really necessary.

**Comment by mitchellbusby** [comment]
Hi @aditya-bhuyan, I tried doing that, and setting it to several other mirrors (such as my local Australian one and the EU one) and it didn't fix the issue. But thank you for the advice :)

**Comment by citychant** [comment]
I am on the same foot, getting nowhere. I have no proxy and tried everything mentioned above, all in vein....

**Comment by cyraid** [comment]
I am with "citychant" on this one.. It's quite annoying that there is no 'resume' when fetching the packages.. There is just start over.

**Comment by wirehack7** [comment]
Having the same problem here. Also ```wget https://registry.npmjs.org/socket.io/-/socket.io-0.9.12.tgz``` fails with timed out. When I download it via TOR it works. Maybe the problem is also at the repo.

**Comment by TimeForAChange** [comment]
@wirehack7 I don't think your case fits into this problem. This issue here has as a core property that connectivity to the destination <i>registry.npmjs.org</i> works without a problem and that <i>only</i> `npm update|install` has the problem.

**Comment by gabor-nagy** [comment]
I tried all the suggested solutions with no luck but finally *disabling my router's firewall* solved the issue. Strange.

**Comment by TFarla** [comment]
@aditya-bhuyan Thanks, That fixed it for me!

**Comment by knownasilya** [comment]
Yes, disabling the belkin firewall did the trick. Thanks all!

**Comment by yadavgn** [comment]
I was struggling with this issue for a long time and npm update worked for me.
used following command to update npm 
npm install npm -g

**Comment by yadavgn** [comment]
Thanks every for the solutions. 
i think two different solutions worked for me. 
update npm using cmd > npm install npm -g
and disabling my router's firewall.

**Comment by marioagarcia** [comment]
This did it for me

`npm set registry https://registry.npmjs.org/`

**Comment by sagaint** [comment]
Me too was facing same issue, but I solved just disabling the windows firewall and McAfee realtime.

**Comment by donPuerto** [comment]
this works for me npm 
### set registry https://registry.npmjs.org/

Thanks mario

**Comment by abernh** [comment]
setting the registry didn't do anything for me.
*using a VPN* solved it for me - so I guess I also had some firewall issues.

**Comment by anandagrawal84** [comment]
setting registry dint work for me either. 

Disabling belkin router firewall worked for me. Strange..

**Comment by Makadas** [comment]
please help, none of the above worked I only have to try the vpn router option. Even the npm website gets blocked.
What do I need to change in my router settings?

**Comment by abernh** [comment]
@Makadas - if even the npm website is blocked you might have a different problem.
But regarding the VPN solution, this is less about configuring your router (unless you have a pretty good, well customizable one) and more about you, using a VPN from the computer, that is going to download from npm.
A free, simple-click-n-happy solution (for <500mb traffic) is offered by the guys from http://tunnelbear.com (register, download, activate, ready, npm). Or you go custom using the tutorial on https://openvpn.net/index.php/open-source/documentation/howto.html

**Comment by LAXITKUMAR** [comment]
This worked for me(not using proxy):

1. set registry mirror for npm..

    npm config set registry http://skimdb.npmjs.com/registry

found mirror from docs:[https://docs.npmjs.com/misc/registry][1]


  [1]: https://docs.npmjs.com/misc/registry

2. after this try npm install if it works then try to install whatever you want otherwise follow below steps as well

3.npm install -g handlebar //i did because it was showing error in npm log but you can skip

4.after that try to set again official registry

    npm config set registry http://registry.npmjs.org

5.now try to install whatever package you want :-)

**Comment by Makadas** [comment]
nothing worked.
So at the end I uninstalled Node and cleared all the caches and started from scratch making sure there were no traces on my computer of npm.
then everything worked like a charm!

I guess with these things...you don't always really know what the problem was

**Comment by barrars** [comment]
Check that your network isnt set to 'public'.  I dont know why mine was but changing it to 'private' seems to have fixed my issue.

**Comment by IngwiePhoenix** [comment]
I have a feeling that the only reason this is happening to me is because of 30+ HTTP connections that are being thrown at NPM.

Installing small things works. But my project has a somewhat large number of deps, plus the ones that sub-dependencies have and alike.

If NPM could "serialize" the download of packages, that would help a ton. But right now, NPM will just...freeze and tmeout.

**Comment by IngwiePhoenix** [comment]
133 dependencies, by the way. That is what `Object.keys(require("./package.json").dependencies).length` gives me.

**Comment by Lughino** [comment]
`npm config set registry http://registry.npmjs.org` worked for me!

**Comment by agalli** [comment]
I spent 5 hours trying to solve this issuse, it was due my belkin router firewall :(

**Comment by enzosaba** [comment]
disabling my router firewall (which is not belkin) did it for me.

**Comment by enzosaba** [comment]
actually no need to disable completely the router firewall. It's enough to disable the function: "SPI and Anti-DoS firewall protection"

**Comment by heystewart** [comment]
"solved" in my case, thanks to @iarna's post and crapify. Here are the details to hopefully help some other poor soul in the future...

- Virtual Box with Ubuntu, on a Mac, with a wireless connection. 
- npm install on the vm would timeout during the initial stages when downloading packages:
```
npm ERR! git fetch -a origin (ssh://git@github.com/xxx.git) ssh: connect to host github.com port 22: Connection timed out
npm ERR! git fetch -a origin (ssh://git@github.com/xxx.git) fatal: Could not read from remote repository.
npm ERR! git fetch -a origin (ssh://git@github.com/xxx.git)
npm ERR! git fetch -a origin (ssh://git@github.com/xxx.git) Please make sure you have the correct access rights
npm ERR! git fetch -a origin (ssh://git@github.com/xxx.git) and the repository exists.
```
- ssh -T git@github.com works, and in fact fetching individual items would work
- failed on two different networks (a corporate one and a personal one)
- only configuration that worked was having a direct ethernet connection and a bridged adapter for virtual box

After reading @iarna's post, I tried using https://github.com/bcoe/crapify:

```
$ crapify start --port=5000 --concurrency=40 --speed=100000000
   (the default for crapify is 10000 for the speed for some reason, rather than no throttling. Oddly, with the default, I would get tar errors...)
$ npm config set proxy http://127.0.0.1:5000 
   (use 'npm config delete proxy' to remove this setting)
$ npm install
```

From the crapify messages, it seems for my project the npm install creates upwards of 500 simultaneous connections, and it appears this is the cause of my issues. For some reason noone else in my group had any problems. Lucky me!

**Comment by jhourlad** [comment]
Confirming that `npm config set registry http://registry.npmjs.org` does the trick.

**Comment by saurshaz** [comment]
If like me - If you're on a VPS, it could be because of low RAM. you can get out of it with higher RAM or turning swap on.
see here `https://www.digitalocean.com/community/tutorials/how-to-add-swap-on-ubuntu-12-04`

**Comment by saurshaz** [comment]
Eventually i got this fix the issue -
`npm@3.8.7` and node `v4.2.0`. I think npm version has the fix.

**Comment by MannikJ** [comment]
I just want to add that my problems started after installing the Cisco AnyConnect VPN Client on Windows 10. Before installation everything was working fine.
Though `npm install` seems to work fine when connected to a VPN, `npm update -g` causes a reconnect of the VPN connection, but also finishes without errors. All in all a very strange and annoying behaviour. 
Uninstallation of AnyConnect did not change anything. Maybe someone has a clue what's going on...

**Comment by khan789** [comment]
Ok, This solved my problem. 
In windows I have Environment variable as NODE_ENV=production. I removed it and NPM started working again for me.

**Comment by khan789** [comment]
Ok, This solved my problem. 
In windows I have Environment variable as NODE_ENV=production. I removed it and NPM started working again for me.

**Comment by michelgotta** [comment]
Maybe another solution: https://github.com/npm/npm/issues/7781#issuecomment-234907353

**Comment by mayerwin** [comment]
Still same problem, none of the suggested "solutions" work. Seriously, something is wrong with npm if you have to disable a router's firewall to make it work...

**Comment by mayerwin** [comment]
I had to also restart my router to make it "work"...

**Comment by ranxero** [comment]
same issue here on WIN 7pro/64: 
npm won't update and ends with ETIMEOUT
fix -> disabling windows firewall helped 100%

**Comment by jehoshua02** [comment]
In my case, I started npm install Friday, forgot about it all weekend, and cancelled it on Monday when I saw that it had made no progress.

I tried @heystewart [solution above](https://github.com/npm/npm/issues/8293#issuecomment-195961283). Here's what I think is happening.

1. npm spins up too many simultaneous requests
2. ISP (in my case Comcast) detects this and "slams the door shut" on it (probably out of good intentions -- think massive DDoS attacks -- but poorly implemented)
3. npm just hangs forever

crapify worked because it intercepts and manages connections to keep them under control.

**Comment by iarna** [comment]
@jehoshua02 You can also set`maxsockets` to a lower value. It defaults to `50`. (eg, try it out with `npm install --maxsockets=10 …` or set it with `npm config set maxsockets 10`) In our experience this is usually caused by the home router getting overloaded. (FWIW, I use Comcast and haven't had any problems.)

**Comment by nikhilshinde57** [comment]
Hi All,

I was also facing the same problem. After googling I found one solution and I tried that one and it's working for me. Hope so it will also work for you.
Just run this command on node console:
npm config set proxy http://proxy.address:proxy.port

:) Happy Coding

**Comment by Hardware-Hacks** [comment]
Just for further help on potential solutions.

I hit the same exact error when I had installed PSAD the previous day for intruder detection. I had neglected to ignore attacks from my local ip 127.0.0.1. After adding in the ignore of the localhost IP and restarting PSAD NPM worked fine again.

**Comment by jehoshua02** [comment]
@iarna having this problem again today. I tried setting `maxsockets 1` and it just sits there, no output. Forever.

So I'm trying crapify again, which I can see is doing something, but now getting an untar error on one of my dependencies 😭. Probably unrelated.
